[Woo POS] [Variable Products] Show error UI from pull-to-refresh or initial load failure

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14695 
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR shows the error UI when loading variations fails with a retry CTA. This reuses the SwiftUI view in the root view with some refactoring to have a view without the "Products" title.

🗒️  Please notice an existing issue on the latest trunk from a recent change, I logged it separately in https://github.com/woocommerce/woocommerce-ios/issues/14860.

This pull request includes several changes to improve error handling and code consistency in the Point of Sale (POS) module of the WooCommerce project. The key changes involve updating error states, renaming methods for clarity, and adding new views for error display.

### Error Handling Improvements:
* [`WooCommerce/Classes/POS/Models/PointOfSaleErrorState.swift`](diffhunk://#diff-15e40514fbb21160e9ccbfdb2e2212b8b36baf552cac769d62861cfaf042ad2aL9-R17): Added new error states for loading variations and updated existing error state methods to reflect these changes. [[1]](diffhunk://#diff-15e40514fbb21160e9ccbfdb2e2212b8b36baf552cac769d62861cfaf042ad2aL9-R17) [[2]](diffhunk://#diff-15e40514fbb21160e9ccbfdb2e2212b8b36baf552cac769d62861cfaf042ad2aL27-R62)
* [`WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift`](diffhunk://#diff-b34ef3931ae78bf4b51520edbe9c11abd6bf36d31dfce4882f3b621c5a7e2af9L99-R99): Updated the error handling for loading variations to use the new error state.

### Method Renaming:
* [`WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift`](diffhunk://#diff-b9d72a7102bb1a5778e338bc3b2046cd3490aa8dc893fc3a05f18cb655ce698cL26-R26): Renamed `reloadItems` to `loadItems` for better clarity and consistency across the codebase, something I forgot in https://github.com/woocommerce/woocommerce-ios/pull/14852. [[1]](diffhunk://#diff-b9d72a7102bb1a5778e338bc3b2046cd3490aa8dc893fc3a05f18cb655ce698cL26-R26) [[2]](diffhunk://#diff-b9d72a7102bb1a5778e338bc3b2046cd3490aa8dc893fc3a05f18cb655ce698cL94-R94)

### New Error Views:
* [`WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleItemListErrorView.swift`](diffhunk://#diff-82429f2bf5e76342d5ecd74acb3049331f1036470b8a014617e06ccc13a5917bR3-L13): Added a new reusable view to display error messages with a retry call-to-action (CTA). [[1]](diffhunk://#diff-82429f2bf5e76342d5ecd74acb3049331f1036470b8a014617e06ccc13a5917bR3-L13) [[2]](diffhunk://#diff-82429f2bf5e76342d5ecd74acb3049331f1036470b8a014617e06ccc13a5917bR41-R43)
* [`WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleItemListFullscreenErrorView.swift`](diffhunk://#diff-3b85911e370cd27313d4bd56043d3c0d36cc7cb04a07663f55858a0845d9e20fR1-R22): Created a fullscreen error view for displaying error messages with a retry CTA, used in the root view.

### Integration of New Views:
* [`WooCommerce/Classes/POS/Presentation/Item Selector/ChildItemList.swift`](diffhunk://#diff-e00b9006e8350a3bbba23158a3c7109cda5523bf7eb46287e08588f69439227bL36-R75): Integrated the new error views into the item list view by updating the state handling and adding preview support. [[1]](diffhunk://#diff-e00b9006e8350a3bbba23158a3c7109cda5523bf7eb46287e08588f69439227bL36-R75) [[2]](diffhunk://#diff-e00b9006e8350a3bbba23158a3c7109cda5523bf7eb46287e08588f69439227bR136-R158)
* [`WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift`](diffhunk://#diff-1069a53acabb24cae728a9d5f67de8766ab4ef606dc97df121c2aac09844fdfcL21-R23): Updated the dashboard view to use the new fullscreen error view. [[1]](diffhunk://#diff-1069a53acabb24cae728a9d5f67de8766ab4ef606dc97df121c2aac09844fdfcL21-R23) [[2]](diffhunk://#diff-1069a53acabb24cae728a9d5f67de8766ab4ef606dc97df121c2aac09844fdfcL72-R72)

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

- Switch to a store eligible for POS, with at least one variable product
- Go to Menu > Point of Sale
- Simulate a network error by going offline, or using a proxy tool to abort the request
- Pull to refresh --> the same fullscreen error UI should be shown as before
- Simulate a success response for the products request
- Tap `Retry` --> after a bit, the items should be shown. 🗒️  Please note an existing issue where the loading state isn't shown right away in https://github.com/woocommerce/woocommerce-ios/issues/14860
- Do the same steps above for a variation item list, tapping on a variable product

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/d04daa4b-491b-402a-808e-e762cdedf28e

products | variations
-- | --
![Simulator Screenshot - iPad Pro 11-inch (M4) - 2025-01-14 at 16 03 25](https://github.com/user-attachments/assets/4fd974d1-b444-4961-a7e5-61ce183e973e) | ![Simulator Screenshot - iPad Pro 11-inch (M4) - 2025-01-14 at 16 04 28](https://github.com/user-attachments/assets/e6d1388f-57f3-42ca-8098-eb773602bba2)


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.14

jaclync committed:
Show error view from loading variations similar to the root view.

jaclync committed:
Rename `reloadItems` to `loadItems` in POS aggregate model for consistency in items controller.

jaclync committed:
Fix PTR retry not working.

jaclync committed:
Nit: add struct level comments.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
14695 Vertically centre child list error on screen

Centring this content on the screen using a ZStack matches the other fullscreen errors better. Centring it in a VStack below the navigation makes the screen look off balance.

joshheald committed:
Fix lint error

joshheald reviewed (approved):
Works well, apart from the known issue of not showing the loading indicators. As long as we can fix that before the next release I think it's fine to merge, and I'm sure we can fix it in time..

I made a change to the display of child list errors; they looked off balance being vertically centred in the space below the heading, so I put them in a ZStack and vertically centred them in the screen's height instead.

![IMG_0279](https://github.com/user-attachments/assets/5fabcba2-0714-4d89-a246-7393555ff3e8)

jaclync commented:
> Works well, apart from the known issue of not showing the loading indicators. As long as we can fix that before the next release I think it's fine to merge, and I'm sure we can fix it in time..

Yea I think a workaround to move it to a separate async task instead of the SwiftUI refresh control async function would probably work. I'm working on this after merging this, and if you're trying out Observation framework and it works we can remove the workaround.

> I made a change to the display of child list errors; they looked off balance being vertically centred in the space below the heading, so I put them in a ZStack and vertically centred them in the screen's height instead.

Thanks for noticing & fixing this! 💟

jaclync commented:
👋 @staskus just sharing a flaky test `TapToPayEducationViewModelTests.secondaryAction_when_about_and_has_previous_tap_to_pay_usage()` from [an earlier build](https://buildkite.com/automattic/woocommerce-ios/builds/27446#019463e1-bc51-47e9-a9ec-1533e8dbe9ab) in this PR.

staskus commented:
@jaclync, thanks, I'll take a look at it!
